3-state-optin part 2

torotil opened the pull request.

torotil committed:
Fix conditional form vs. empty rule values.

The webform conditional form is unable to deal with empty forms so we
need a different representation for “No change“.

torotil committed:
Introduce a fourth value "no selection" for not selected radios.

katzmo committed:
[campaignion_newsletters] consolidate values and make them more explicit

katzmo committed:
[campaignion_newsletters] coding style

katzmo committed:
[campaignion_newsletters] Refactoring

Encapsulate prefix function in a class of it’s own.

katzmo committed:
[campaignion_newsletters] Migrations

* webform submitted data →  new prefixed values
* webform conditional rules →  new prefixed values

katzmo committed:
[campaignion_newsletters] Improve conditional UX

Show only rule values that match the current newsletter component.

torotil commented on campaignion_newsletters/tests/ConditionalTest.php:
```diff
@@ -20,44 +20,63 @@ public function setUp() {
+    $this->assertTrue($ne(['no-change'], 'radios:not-selected', $radios));
+    $this->assertTrue($ne(['no-change'], 'checkbox:opt-in', $radios));
+    $this->assertTrue($ne(['no-change'], 'checkbox:no-change', $radios));
+    $this->assertTrue($ne(['not-selected'], 'radios:opt-in', $radios));
```
This is the test-case for testing the values directly provided by the form-API. So I guess it should still be a `[]` instead of `['not-selected']`.

torotil replied:
(Fixed that myself.)

katzmo replied:
Your right! But this applies to the second block, testing what doesn't match, too, doesn't it?

torotil replied:
Yes, you’re right too :-)

torotil reviewed (commented):
Almost there I think.

torotil committed:
Fix test for newsletter radio conditionals.

torotil committed:
Make form submission behave according to the new values.

torotil committed:
Fix handling of the default value.

torotil committed:
[campaignion_newsletters] Add migration for default values.

torotil committed:
ValuePrefix::remove() usable with strings too (for the default value).

torotil committed:
New newsletter values → thank you page redirect.

torotil committed:
Conditional redirects: New newsletter values.

torotil committed:
Fixup migration with flipped old/new values.

torotil committed:
Fix the rest of the radio test values in ConditionalTest.php

torotil committed:
Redo conditional redirect opt-in filter for opt-outs.

- Interactions in the form always precede over existing opt-ins.
- An opt-out makes the filter evaluate as “no opt-in” regardless of
  still existing subscriptions.

torotil committed:
Add test for conditional form callback.
